dcc-import-gene/src/main/java/org/icgc/dcc/imports/gene/writer/GeneWriter.java

```diff
@@ -17,139 +17,478 @@
+    val type = line[2].trim();
+    String locStart = line[3].trim();
+    String locEnd = line[4].trim();
+
+    Double score;
+    try {
+      score = Double.parseDouble(line[5].trim());
+    } catch (Exception var15) {
+      score = 0.0D;
+    }
+
+    char strand = line[6].trim().charAt(0);
+    int locationStart = Integer.parseInt(locStart);
+    int locationEnd = Integer.parseInt(locEnd);
+    if (locationStart > locationEnd) {
+      int location = locationStart;
+      locationStart = locationEnd;
+      locationEnd = location;
+    }
+
+    val negative = locationStart <= 0 && locationEnd <= 0;
+    assert strand == 45 == negative;
+
+    int frame;
+    try {
+      frame = Integer.parseInt(line[7].trim());
+    } catch (Exception var14) {
+      frame = -1;
+    }
+
+    val attributes = line[8];
+    val attributeMap = parseAttributes(attributes);
+
+    ObjectNode feature = MAPPER.createObjectNode();
 
-    // Open BSON file stream
-    val genes = readGenes(geneStream);
+    int strandNumber = 0;
+    if (strand == '+') {
+      strandNumber = 1;
+    } else if (strand == '-') {
+      strandNumber = -1;
+    }
+
+    feature.put("seqname", seqname);
+    feature.put("source", source);
+    feature.put("type", type);
+    feature.put("locationStart", locationStart);
+    feature.put("locationEnd", locationEnd);
+    feature.put("score", score);
+    feature.put("strand", strandNumber);
+    feature.put("frame", frame);
 
-    writeGenes(watch, genes, geneFilter);
+    for (val kv : attributeMap.entrySet()) {
+      feature.put(kv.getKey(), kv.getValue());
+    }
+
+    return feature;
+  }
+
+  private String getName(String symbol) {
+    if (nameMap.containsKey(symbol)) {
+      return nameMap.get(symbol);
+    } else {
+      return "";
+    }
   }
 
-  private void writeGenes(Stopwatch watch, MappingIterator<JsonNode> genes, GeneFilter geneFilter)
-      throws UnknownHostException, IOException {
-    val genesCollection = getCollection(GENE_COLLECTION);
+  private ArrayNode getDescription(String id) {
+    if (synMap.containsKey(id)) {
+      return synMap.get(id);
+    } else {
+      return MAPPER.createArrayNode();
+    }
+
+  }
 
-    // Drop the current collection
-    log.info("Dropping current gene collection, if any...");
-    clearGenes(genesCollection);
+  private static HashMap<String, String> parseAttributes(String attributes) {
+    val attributeMap = new HashMap<String, String>();
 
-    log.info("Saving gene documents");
-    saveGenes(genesCollection, genes, geneFilter);
+    String[] tokens = attributes.split(";");
+    for (val token : tokens) {
+      String[] kv = token.trim().replace("\"", "").split("\\s+");
+      attributeMap.put(kv[0], kv[1]);
+    }
+    return attributeMap;
+  }
 
-    log.info("Wrote gene model in {}", watch);
+  private static ObjectNode constructGeneNode(ObjectNode data) {
+    val gene = MAPPER.createObjectNode();
+    gene.put("_gene_id", data.get("gene_id").asText());
```

Comment:
If you are doing this a lot in code, I find adding some utilities to do things like `textValue(data, "gene_id")` really helps to cut down on the noise.